src/dpos/dpos.ts

```diff
@@ -0,0 +1,319 @@
+    console.log('Initialized', dappchainEndpoint)
+    client.txMiddleware = [
+      new NonceTxMiddleware(publicKey, client),
+      new SignedTxMiddleware(privateKey)
+    ]
+
+    client.on('error', (msg: any) => {
+      console.error('PlasmaChain connection error', msg)
+    })
+
+    const address = new Address(chainId, LocalAddress.fromPublicKey(publicKey))
+
+    const dappchainLoom = await Coin.createAsync(client, address)
+    const dappchainDPOS = await DPOS2.createAsync(client, address)
+    const dappchainGateway = await LoomCoinTransferGateway.createAsync(client, address)
+    await this.mapAccounts(client, address, wallet)
+    return new DPOSUser(
+      wallet,
+      client,
+      address,
+      gatewayAddress,
+      loomAddress,
+      dappchainGateway,
+      dappchainLoom,
+      dappchainDPOS
+    )
+  }
+
+  /**
+   * Maps the user's ETH address to their DAppChain address. This MUST be called before any interaction with the gateways.
+   *
+   * @param account The user's account object
+   * @param wallet The User's ethers wallet
+   */
+  static mapAccounts = async (client: Client, address: Address, wallet: ethers.Signer) => {
+    const walletAddress = await wallet.getAddress()
+    const ethereumAddress = Address.fromString(`eth:${walletAddress}`)
+    const mapperContract = await AddressMapper.createAsync(client, address)
+    if (await mapperContract.hasMappingAsync(address)) {
+      console.error(`${address.toString()} is already mapped`)
+      return
+    }
+    const signer = new EthersSigner(wallet)
+    await mapperContract.addIdentityMappingAsync(address, ethereumAddress, signer)
+    console.log(`Mapped ${address} to ${ethereumAddress}`)
+  }
+
+  constructor(
+    wallet: ethers.Signer,
+    client: Client,
+    address: Address,
+    gatewayAddress: string,
+    loomAddress: string,
+    dappchainGateway: Contracts.LoomCoinTransferGateway,
+    dappchainLoom: Contracts.Coin,
+    dappchainDPOS: Contracts.DPOS2
+  ) {
+    this._wallet = wallet
+    this._address = address
+    this._client = client
+    this._ethereumGateway = new ethers.Contract(gatewayAddress, ERC20Gateway, wallet)
+    this._ethereumLoom = new ethers.Contract(loomAddress, ERC20, wallet)
+    this._dappchainGateway = dappchainGateway
+    this._dappchainLoom = dappchainLoom
+    this._dappchainDPOS = dappchainDPOS
+  }
+
+  listValidators = async (): Promise<IValidator[]> => {
+    return this._dappchainDPOS.getValidatorsAsync()
+  }
+
+  listCandidates = async (): Promise<ICandidate[]> => {
+    return this._dappchainDPOS.getCandidatesAsync()
+  }
+
+  /**
+   * Deposits funds from mainnet to the gateway
+   */
+  deposit = async (amount: BN): Promise<ethers.ContractTransaction> => {
+    let currentApproval = await this._ethereumLoom.allowance(
+      await this._wallet.getAddress(),
+      this._ethereumGateway.address
+    )
+    currentApproval = new BN(currentApproval._hex.split('0x')[1], 16) // ugly way to convert the BN
```

Comment:
Just `new BN(currentApproval._hex, 16)` doesn't work?

Reply:
Nah. Some weird stuff happening because the BN returned by Ethers is actually different from the BNs in BN.js. I'll take care of this sometime in the future, but let's merge it for now